actix-server/src/server.rs

```diff
@@ -228,67 +206,63 @@ impl ServerInner {
 
         let (waker_queue, worker_handles) = Accept::start(sockets, &builder)?;
 
-        // construct OS signals listener future
-        let signals = (builder.listen_os_signals).then(Signals::new);
-
-        Ok(ServerInner {
-            cmd_tx: builder.cmd_tx.clone(),
+        let mux = ServerEventMultiplexer {
+            signal_fut: (builder.listen_os_signals).then(Signals::new),
             cmd_rx: builder.cmd_rx,
-            signals,
+        };
+
+        let server = ServerInner {
             waker_queue,
             worker_handles,
             worker_config: builder.worker_config,
             services: builder.factories,
             exit: builder.exit,
-            stop_task: None,
-        })
+            stopped: false,
+        };
+
+        Ok((server, mux))
     }
 
-    fn handle_cmd(&mut self, item: ServerCommand) -> Option<BoxFuture<'static, ()>> {
+    async fn handle_cmd(&mut self, item: ServerCommand) {
         match item {
             ServerCommand::Pause(tx) => {
                 self.waker_queue.wake(WakerInterest::Pause);
                 let _ = tx.send(());
-                None
             }
 
             ServerCommand::Resume(tx) => {
                 self.waker_queue.wake(WakerInterest::Resume);
                 let _ = tx.send(());
-                None
             }
 
             ServerCommand::Stop {
                 graceful,
                 completion,
+                force_exit,
             } => {
-                let exit = self.exit;
+                self.stopped = true;
 
                 // stop accept thread
                 self.waker_queue.wake(WakerInterest::Stop);
 
-                // stop workers
-                let workers_stop = self
-                    .worker_handles
-                    .iter()
-                    .map(|worker| worker.stop(graceful))
-                    .collect::<Vec<_>>();
-
-                Some(Box::pin(async move {
-                    if graceful {
-                        // wait for all workers to shut down
-                        let _ = join_all(workers_stop).await;
-                    }
+                if graceful {
+                    // wait for all workers to shut down
```

Comment:
do we not want to send the stop messages even in the case of a force shutdown ?

Reply:
oh, right!